astropy/modeling/physical_models.py

```diff
@@ -371,3 +372,334 @@ def input_units(self):
+        if masstype == "virial":
+            Om_c = cosmo.Om(redshift) - 1.0
+            d_c = 18.0 * np.pi ** 2 + 82.0 * Om_c - 39.0 * Om_c ** 2
+            self.density_delta = d_c * cosmo.critical_density(redshift)
+        elif masstype == 'c':
+            self.density_delta = delta * cosmo.critical_density(redshift)
+        elif masstype == 'm':
+            self.density_delta = delta * cosmo.critical_density(redshift) * cosmo.Om(redshift)
+        else:
+            raise ValueError("Invalid masstype '" + str(masstype) +
+                             "'. Should be one of 'virial','c', or 'm'")
+
+        # Set scale radius and density
+        # Delta Mass is related to delta radius by
+        # M_{200}=\frac{4}{3}\pi r_{200}^3 200 \rho_{c}
+        # And delta radius is related to the NFW scale radius by
+        # c = R / r_{\\rm s}
+
+        # Establish mass units for density calculation (default solar masses)
+        if not isinstance(mass, u.Quantity):
+            in_mass = u.Quantity(mass, u.M_sun)
+        else:
+            in_mass = mass
+
+        # Calculate scale radius
+        # Set radial units to kiloparsec by default (unit will be rescaled by units of radius
+        # in evaluate)
+        self.radius_s = (((3.0 * in_mass) / (4.0 * np.pi * self.density_delta)) ** (
+                    1.0 / 3.0)) / concentration
+        self.radius_s = self.radius_s.to(u.kpc)
+
+        # Calculate scale density
+        self.density_s = in_mass / (4.0 * np.pi * self.radius_s ** 3 * self.A_NFW(concentration))
+
+        super().__init__(mass=in_mass, concentration=concentration, redshift=redshift, **kwargs)
+
+    def evaluate(self, r, mass, concentration, redshift):
+        """
+        One dimensional NFW profile function
+
+        Parameters
+        ----------
+        r : float or :class:`~astropy.units.Quantity`
+            Radial position of density to be calculated for the NFW profile.
+        mass : float or :class:`~astropy.units.Quantity`
+            Mass of NFW peak within specified overdensity radius.
+        concentration : float
+            Concentration of the NFW profile.
+        redshift : float
+            Redshift of the NFW profile.
+
+        Returns
+        -------
+        density : float or :class:`~astropy.units.Quantity`
+            NFW profile mass density at location ``r``. The density units are:
+            [``mass`` / ``r`` ^3]
+
+        Notes
+        -----
+        .. warning::
+
+            Output values might contain ``nan`` and ``inf``.
+        """
+        # Enforce default units
+        if not isinstance(mass, u.Quantity):
+            in_mass = u.Quantity(mass, u.M_sun)
+        else:
+            in_mass = mass
+
+        # Create radial version of input with dimension
+        if hasattr(r, "unit"):
+            in_r = r
+            self.radius_s = self.radius_s.to(r.unit)
+            self.density_s = self.density_s.to(in_mass.unit / r.unit**3)
```

Comment:
Why are `radius_s` and `denstiy_s` computed here and further down?

Reply:
This block is labeled with:
# Update NFW Scale radius
...
# Update NFW Scale density

To highlight that these values are being updated. The reason this update code exists is in the case a user updates the parameters in their evaluate call from what they gave in the init (i.e. like the fitter does). Accordingly, these parameters would need to also be updated.

Reply:
Private function added as requested.